<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=iso-8859-1">
<title>libexcept: Limitations</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.5.1 -->
<div class="tabs">
  <ul>
    <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
    <li><a href="modules.html"><span>Modules</span></a></li>
    <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
  </ul></div>
<h1><a class="anchor" name="limits">Limitations</a></h1>Although it is possible to get a fairly functional exceptions framework using standard ANSI C, there are limitations to the system. This page documents things that cannot be done with libexcept.<h2><a class="anchor" name="breakout">
Breaking out of an exception block</a></h2>
It is not possible to break out of an exception block using continue, break, return or goto. libexcept has built-in guards that detect when this is done. The following example demonstrates illegal usage:<p>
<div class="fragment"><pre class="fragment">    <span class="keywordtype">void</span> illegal(<span class="keywordtype">void</span>)
    {
        <span class="keywordtype">int</span> i;

        <span class="keywordflow">for</span> (i=0; i&lt;10; ++i) {
            except_try {
                <span class="comment">/* Illegal: cannot use break to exit out of an exception block */</span>

                <span class="keywordflow">break</span>;
            
                <span class="comment">/* Illegal: cannot use continue to exit out of an exception block */</span>

                <span class="keywordflow">continue</span>;

                <span class="comment">/* Illegal: cannot return from within an exception block */</span>

                <span class="keywordflow">return</span>;

                <span class="comment">/* Illegal: do not use goto to break out of an exception block */</span>

                <span class="keywordflow">goto</span> label;
            } except_end

        label:
        }
    }
</pre></div><p>
However, the following is legal:<p>
<div class="fragment"><pre class="fragment">    <span class="keywordtype">void</span> legal(<span class="keywordtype">void</span>)
    {
        <span class="keywordtype">int</span> i;

        except_try {
            <span class="keywordflow">for</span> (i=0; i&lt;10; ++i) {
                <span class="comment">/* Legal: Can use break to exit a loop within an exception block */</span>

                <span class="keywordflow">break</span>;
            
                <span class="comment">/* Legal: Can use continue within an exception block */</span>

                <span class="keywordflow">continue</span>;

                <span class="comment">/* Legal: Can use goto to jump within an exception block */</span>

                <span class="keywordflow">goto</span> label;
            }
            label:
        } except_end
    }
</pre></div><p>
Note: the same limitations also apply within catch and finally blocks.<h2><a class="anchor" name="nested">
Nested exception blocks</a></h2>
It is not possible to have an exception block within another exception block. The following is illegal:<p>
<div class="fragment"><pre class="fragment">    <span class="keywordtype">void</span> illegal(<span class="keywordtype">void</span>)
    {
        except_try {
            except_try {
                printf(<span class="stringliteral">"hello, world\n"</span>);
            } except_catch (CircleException, e) {
                printf(<span class="stringliteral">"Circle Exception caught!\n"</span>);
            }
        } except_catch (SquareException, e) {
            printf(<span class="stringliteral">"Square Exception caught!\n"</span>);
        }
    }
</pre></div><p>
However, the following is legal:<p>
<div class="fragment"><pre class="fragment">    <span class="keywordtype">void</span> legal_2(<span class="keywordtype">void</span>)
    {
        except_try {
            printf(<span class="stringliteral">"hello, world\n"</span>);
        } except_catch (CircleException, e) {
            printf(<span class="stringliteral">"Circle Exception caught!\n"</span>);
        }
    }

    <span class="keywordtype">void</span> legal_1(<span class="keywordtype">void</span>)
    {
        except_try {
            legal_2();
        } except_catch (SquareException, e) {
            printf(<span class="stringliteral">"Square Exception caught!\n"</span>);
        }
    }
</pre></div> <hr size="1"><address style="align: right;"><small>Generated on Sun Sep 9 21:35:15 2007 for libexcept by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.5.1 </small></address>
</body>
</html>
